File: assignment/assignment1/scala/hw1/src/main/scala/task3.scala
/*
Scala: $ spark-submit --class task3 hw1.jar <input_file> <output_file> <partition_type> <n_partitions> <n>

$ sbt package
// default partitioner
$ spark-submit --class task3 ./target/scala-2.11/hw1_2.11-0.1-SNAPSHOT.jar "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json" "output_task3_scala.json" "default" 30 150
// customized partitioner
$ spark-submit --class task3 ./target/scala-2.11/hw1_2.11-0.1-SNAPSHOT.jar "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json" "output_task3_scala.json" "customized" 30 150
*/

import java.io.PrintWriter
import org.apache.spark.SparkContext
import org.apache.spark.SparkContext._
import org.apache.spark.SparkConf
import org.apache.spark.Partitioner

import org.json4s._
import org.json4s.jackson.JsonMethods._
import org.json4s.jackson.Serialization

class BusinessidPartitioner(numParts:Int) extends Partitioner{
    override def numPartitions: Int = numParts

    override def getPartition(key: Any): Int = {
        key.toString.hashCode % numPartitions
    }

    override def equals(other: Any): Boolean = other match {
		case t: BusinessidPartitioner => t.numPartitions == numPartitions
		case _ => false
	}
}

object task3 {
    def main(args: Array[String]) {
        // parameters
        val input_file = args(0)
        val output_file = args(1)
        val partition_type = args(2)
        val n_partitions = args(3).toInt
        val n = args(4).toInt

        // val input_file =  "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json"
        // val stopwords_path = "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/stopwords"

        val conf = new SparkConf().setAppName("task3").setMaster("local[*]")
        val sc = new SparkContext(conf)

        def textToMapBusinessId(r: String): (String, Int) = {
            implicit val formats = DefaultFormats
            val m = parse(r).extract[Map[String, Any]]
            (m("business_id").toString, 1)
        }


        var business_id_data = sc.textFile(input_file).map(textToMapBusinessId)

        if (partition_type == "default") {
            business_id_data = business_id_data.cache()
        } else if (partition_type == "customized") {
            business_id_data = business_id_data.partitionBy(new BusinessidPartitioner(n_partitions)).cache()
        } else {
            printf("Wrong partition type.")
        }
        
        // println(business_id_data.getClass)

        // number_of_partitions
        val number_of_partitions = business_id_data.getNumPartitions
        println(number_of_partitions)

        // number_of_items_in_each_partition
        def countItemsInAPartition(iter: Iterator[Any]): Iterator[Int] = {
            val count = List[Int]()
            val i = iter.length
            count.::(i).iterator
        }
        val number_of_items_in_each_partition = business_id_data.mapPartitions(countItemsInAPartition).collect
        number_of_items_in_each_partition.foreach(println)

        // businesses_that_have_more_than_n_reviews
        val businesses_that_have_more_than_n_reviews = business_id_data.reduceByKey(_+_)
                                                                    .filter(_._2 > n)
                                                                    .map(x => List(x._1.toString, x._2))
                                                                    .collect
        businesses_that_have_more_than_n_reviews.foreach(println)

        // result
        val result = Map("n_partitions"->number_of_partitions, "n_items"->number_of_items_in_each_partition, "result"->businesses_that_have_more_than_n_reviews)
        implicit val formats = DefaultFormats
        val result_json = Serialization.write(result)
        println(result_json)

        // write to output_task3_scala.json file
        val out = new PrintWriter(output_file)
        out.print(result_json)
        out.close()

    }
}

File: assignment/assignment1/python/task3.py
"""
Python: $ spark-submit task3.py <input_file> <output_file> <partition_type> <n_partitions> <n> 
Params:
input_file – the input file (the review dataset) 
output_file – the output file contains your answers 
partition_type – the partition function, either “default” or “customized” 
n_partitions – the number of partitions (only effective for the customized partition function) 
n – the threshold of the number of reviews (see 4.3.1)

default:
spark-submit task3.py "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json" "output_task3.json" "default" 50 150 
customized:
spark-submit task3.py "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json" "output_task3.json" "customized" 50 150 
"""

import sys
import json
import time
from pyspark import SparkContext, SparkConf

# input_file = "file:///Users/markduan/duan/USC_course/USC_APDS/INF553/homework/hw1/hw1_data/review.json"
# output_file = "output_task3.json"
# partition_type = "customized" # either "default" or "customized"
# n_partitions = 30 
# n = 150

input_file = sys.argv[1]
output_file = sys.argv[2]
partition_type = sys.argv[3]
n_partitions = int(sys.argv[4])
n = int(sys.argv[5])

def timer(func):
    """timer decorator"""
    def wrapper(*arg, **karg):
        t1 = time.time()
        res = func(*arg, **karg)
        t2 = time.time()
        delta_t = t2 - t1
        print('time consume:%.2fs' % delta_t)
        return res
    return wrapper

def getNumberOfItems(iter):
    par = list(iter)
    n_par = len(par)
    yield n_par

def write_output_file(result, output_file):
    # write on local disk instead of hdfs
    with open(output_file, 'w', encoding='utf-8') as fp:
        json.dump(result, fp)

def customPartitionFunc(business_id):
    return hash(business_id)

@timer
def main(n_p='default'):
    output = {}

    # define spark env
    conf = SparkConf() \
        .setAppName("task3") \
        .setMaster("local[*]")
    sc = SparkContext(conf=conf)

    review_file_path = input_file
    review_data = None
    if n_p == 'default':
        review_data = sc.textFile(review_file_path) \
            .map(lambda r: [json.loads(r)['business_id'], 1]) \
            .cache()
    else:
        review_data = sc.textFile(review_file_path) \
            .map(lambda r: [json.loads(r)['business_id'], 1]) \
            .partitionBy(n_p, partitionFunc=customPartitionFunc) \
            .cache()

    # get n_partitions
    n_partitions = review_data.getNumPartitions()
    output['n_partitions'] = n_partitions

    # get n_items
    n_items = review_data.mapPartitions(getNumberOfItems).collect()
    output['n_items'] = n_items

    # get result
    businesses_that_have_more_than_n_reviews = review_data \
        .reduceByKey(lambda x, y: x+y) \
        .filter(lambda x: x[1] > n) \
        .map(lambda x: [x[0], x[1]]) \
        .collect()
    output['result'] = businesses_that_have_more_than_n_reviews

    write_output_file(output, output_file)

if __name__ == "__main__":
    if partition_type == "default":
        main()
    elif partition_type == "customized":
        main(n_partitions)
    else:
        print('wrong partition_type value.')

